From 5a7ba9b7b0a7d2caa20ed0a276c31a99a34fcfad Mon Sep 17 00:00:00 2001
From: Mooneer Salem <mooneer@gmail.com>
Date: Fri, 17 Sep 2021 22:46:24 -0700
Subject: [PATCH 08/53] Only override printf for SM1000.

---
 src/dump.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/dump.c b/src/dump.c
index 4fae3b67..0a19943e 100644
--- a/src/dump.c
+++ b/src/dump.c
@@ -32,7 +32,7 @@
 #include <string.h>
 #include <math.h>
 
-#ifdef __EMBEDDED__
+#ifdef STM32F40_41xxx
 #include "gdb_stdio.h"
 #define fprintf gdb_stdio_fprintf
 #define fopen gdb_stdio_fopen
-- 
2.30.2

