From bd9f298ceae465c16f19a1986da35077415e7937 Mon Sep 17 00:00:00 2001
From: Mooneer Salem <mooneer@gmail.com>
Date: Mon, 27 Sep 2021 18:37:23 -0700
Subject: [PATCH 11/53] Revert changes to ofdm.c due to need for faster ARM
 math algorithms.

---
 src/ofdm.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/src/ofdm.c b/src/ofdm.c
index 165d9bc4..a6a404e5 100644
--- a/src/ofdm.c
+++ b/src/ofdm.c
@@ -43,9 +43,9 @@
 #include "debug_alloc.h"
 #include "machdep.h"
 
-#ifdef STM32F40_41xxx
+#ifdef __EMBEDDED__
 #include "arm_math.h"
-#endif /* STM32F40_41xxx */
+#endif /* __EMBEDDED__ */
 
 /* Static Prototypes */
 
@@ -731,7 +731,7 @@ static int est_timing(struct OFDM *ofdm, complex float *rx, int length,
         corr_st = 0.0f;
         corr_en = 0.0f;
 
-#ifdef STM32F40_41xxx
+#ifdef __EMBEDDED__
 #ifdef __REAL__
 	float re,im;
 
@@ -758,7 +758,7 @@ static int est_timing(struct OFDM *ofdm, complex float *rx, int length,
 	    corr_st = corr_st + (rx[ind                        ] * wvec_pilot[j]);
             corr_en = corr_en + (rx[ind + ofdm->samplesperframe] * wvec_pilot[j]);
         }
-#endif // STM32F40_41xxx
+#endif // __EMBEDDED__
         corr[i] = (cabsf(corr_st) + cabsf(corr_en)) * av_level;
     }
 
-- 
2.30.2

