From 2ab504b9559b7eb25f4cee019ba4dce87fa249f4 Mon Sep 17 00:00:00 2001
From: Mooneer Salem <mooneer@gmail.com>
Date: Sun, 10 Oct 2021 02:53:53 -0700
Subject: [PATCH 12/53] Update required file headers for FDV_ARM_MATH.

---
 src/codec2_fft.h | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/src/codec2_fft.h b/src/codec2_fft.h
index c7412024..66e8ca2c 100644
--- a/src/codec2_fft.h
+++ b/src/codec2_fft.h
@@ -14,10 +14,11 @@
 #include <string.h>
 #include <math.h>
 
-#ifdef FDV_ARM_MATH
-  #include "fdv_arm_math.h"
+#ifndef FDV_ARM_MATH
+  #define USE_KISS_FFT
 #else
-    #define USE_KISS_FFT
+  #include "arm_math.h"
+  #include "arm_const_structs.h"
 #endif
 
 #include "defines.h"
-- 
2.30.2

