From 86a41680eb08967f11a830473d457857427d01ad Mon Sep 17 00:00:00 2001
From: David Rowe <david@rowetel.com>
Date: Sun, 31 Oct 2021 07:35:43 +1030
Subject: [PATCH 37/53] cleaned up a little

---
 stm32/CMakeLists.txt | 18 +-----------------
 1 file changed, 1 insertion(+), 17 deletions(-)

diff --git a/stm32/CMakeLists.txt b/stm32/CMakeLists.txt
index 10c821ec..82253928 100644
--- a/stm32/CMakeLists.txt
+++ b/stm32/CMakeLists.txt
@@ -1,5 +1,5 @@
 #
-# stm32f4 Codec2 test programs
+# stm32f4 support for Codec2
 #
 # CMake configuration contributed by Richard Shaw (KF5OIM)
 # Please report questions, comments, problems, or patches to the freetel
@@ -81,15 +81,6 @@ if(VERIFY_OPT)
     add_definitions(-DDUMP)
 endif()
 
-#option(SEMIHOSTING "Enable for standard arm semihosting." OFF)
-#if(SEMIHOSTING)
-#    link_libraries(rdimon)
-#    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} --specs=rdimon.specs")
-#else()
-#    link_libraries(nosys)
-#    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -specs=nosys.specs")
-#endif()
-
 # Set default build type
 if(NOT CMAKE_BUILD_TYPE)
     set(CMAKE_BUILD_TYPE "Debug")
@@ -107,13 +98,11 @@ set(CMAKE_EXECUTABLE_SUFFIX_CXX ".elf")
 set(CMAKE_EXECUTABLE_SUFFIX_ASM ".elf")
 set(CMAKE_ASM_FLAGS "${CMAKE_C_FLAGS} -x assembler-with-cpp")
 
-
 # Check build flags
 
 message(STATUS "CMAKE_C_FLAGS: " ${CMAKE_C_FLAGS})
 message(STATUS "CMAKE_ASM_FLAGS: " ${CMAKE_ASM_FLAGS})
 
-
 ###################################################
 
 # STM32F4 Standard Peripheral Library
@@ -158,8 +147,6 @@ include_directories(../src ../unittest inc ${PROJECT_BINARY_DIR})
 
 add_subdirectory(unittest/src)
 
-
-
 # Codec 2
 
 # Output path is such that #include <codec2/version.h> in codec2.h works
@@ -332,6 +319,3 @@ target_link_libraries(sm1000v5 stm32f4_adac stm32f4 CMSIS)
 target_compile_options(sm1000v5 PRIVATE "-O3")
 elf2bin(sm1000v5)
 elf2dfu(sm1000v5)
-
-
-
-- 
2.30.2

